core/src/abstractions.rs

```diff
@@ -11,102 +11,150 @@ use std::{
-        let res = self.sem.clone().acquire_owned().await?;
+    pub(crate) async fn acquire_owned(&self) -> Result<OwnedMeteredSemPermit<SK>, ()> {
+        if let Some(max) = self.max_permits {
+            self.extant_permits
+                .1
+                .clone()
+                .wait_for(|&ep| ep < max)
+                .await
+                .expect("Extant permit channel is never closed");
+        }
+        let res = self.supplier.reserve_slot(self).await;
         Ok(self.build_owned(res))
     }
 
-    pub(crate) fn try_acquire_owned(&self) -> Result<OwnedMeteredSemPermit, TryAcquireError> {
-        let res = self.sem.clone().try_acquire_owned()?;
-        Ok(self.build_owned(res))
+    pub(crate) fn try_acquire_owned(&self) -> Result<OwnedMeteredSemPermit<SK>, ()> {
+        if let Some(max) = self.max_permits {
+            if *self.extant_permits.1.borrow() >= max {
+                return Err(());
+            }
+        }
+        if let Some(res) = self.supplier.try_reserve_slot(self) {
+            Ok(self.build_owned(res))
+        } else {
+            Err(())
+        }
     }
 
-    fn build_owned(&self, res: OwnedSemaphorePermit) -> OwnedMeteredSemPermit {
+    fn build_owned(&self, res: SlotSupplierPermit) -> OwnedMeteredSemPermit<SK> {
         self.unused_claimants.fetch_add(1, Ordering::Release);
-        self.record();
+        self.extant_permits.0.send_modify(|ep| *ep += 1);
+        // Eww
+        let uc_c = self.unused_claimants.clone();
+        let ep_rx_c = self.extant_permits.1.clone();
+        let ep_tx_c = self.extant_permits.0.clone();
+        let supp = self.supplier.clone();
+        let supp_c = self.supplier.clone();
+        let supp_c_c = self.supplier.clone();
+        let mets = self.metrics_ctx.clone();
+        let metric_rec =
+            // When being called from the drop impl, the permit isn't actually dropped yet, so
+            // account for that with the `add_one` parameter.
+            move |add_one: bool| {
+                let extra = usize::from(add_one);
+                let unused = uc_c.load(Ordering::Acquire);
+                if let Some(avail) = supp.available_slots() {
+                    mets.available_task_slots(avail + unused + extra);
+                }
+                mets.task_slots_used((*ep_rx_c.borrow() - unused + extra) as u64);
+            };
+        let mrc = metric_rec.clone();
+        mrc(false);
+
         OwnedMeteredSemPermit {
-            inner: res,
+            _inner: res,
             unused_claimants: Some(self.unused_claimants.clone()),
-            record_fn: self.record_owned(),
+            use_fn: Box::new(move |info| {
+                supp_c.mark_slot_used(info);
+                metric_rec(false)
+            }),
+            release_fn: Box::new(move || {
+                // TODO: Real release reason
+                supp_c_c.release_slot(SlotReleaseReason::TaskComplete);
```

Comment:
This turns out to be... quite annoying to deal with. Since permits simply being `drop`ed causes them to be released, it's incredibly annoying to find every place that might ever happen and attach a reason.

My inclination is to simply not bother with this for now, possibly never, unless users ask for it.

Reply:
I can, but it's worse from a safety perspective (could still use it as a backup though, with some default reason), and it can potentially happen in many, many places. It's just a lot of gruntwork for something I don't know that users will actually care about yet. Hence my preference to just defer it for now.

Reply:
:+1: Can we just get rid of reason for now then?

Reply:
should we remove this from Java then?

Reply:
Yeah probably for consistencies' sake